src/Components/Server/src/Builder/ComponentEndpointRouteBuilderExtensions.cs

```diff
@@ -113,6 +115,33 @@ public static ComponentEndpointConventionBuilder MapBlazorHub(this IEndpointRout
             endpoints.CreateApplicationBuilder().UseMiddleware<CircuitJavaScriptInitializationMiddleware>().Build())
             .WithDisplayName("Blazor initializers");
 
-        return new ComponentEndpointConventionBuilder(hubEndpoint, disconnectEndpoint, jsInitializersEndpoint);
+        var blazorEndpoint = GetBlazorEndpoint(endpoints);
+
+        return new ComponentEndpointConventionBuilder(hubEndpoint, disconnectEndpoint, jsInitializersEndpoint, blazorEndpoint);
+    }
+
+    private static IEndpointConventionBuilder GetBlazorEndpoint(IEndpointRouteBuilder endpoints)
+    {
+        var options = new StaticFileOptions
+        {
+            FileProvider = new ManifestEmbeddedFileProvider(typeof(ComponentEndpointRouteBuilderExtensions).Assembly)
+        };
+
+        var app = endpoints.CreateApplicationBuilder();
+        app.Use(next => context =>
+        {
+            // Set endpoint to null so the static files middleware will handle the request.
+            context.SetEndpoint(null);
+
+            return next(context);
+        });
+        app.UseStaticFiles(options);
+
+        var blazorEndpoint = endpoints.Map("/_framework/blazor.server.js", app.Build())
+            .WithDisplayName("Blazor static files");
+
+        blazorEndpoint.Add((builder) => ((RouteEndpointBuilder)builder).Order = -1);
```

Comment:
The `-1` here makes it so that this endpoint has priority over other endpoints independent of their specificity, so this means that a controller or a file are not going to match over this endpoint. 

With that said, I think we should be a bit more conservative here as people might also register their own routes with a custom order, which might cause those routes to match over this one. For that reason, we should do something similar but opposite to what we do in `MapFallbackToPage` and give it the lowest order possible `(int.MinValue)` so that we can guarantee under most circumstances, no other route will match over this one.

Reply:
The only time this might cause issues is if someone customizes routing extensively, which I am not too worried about.